precompiles/balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+
+			ApprovesStorage::<Runtime, Instance>::get(owner, spender)
+				.unwrap_or_default()
+				.into()
+		};
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: EvmDataWriter::new().write(amount).build(),
+			logs: vec![],
+		})
+	}
+
+	fn approve(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let spender: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		// Write into storage.
+		{
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let spender: Runtime::AccountId = Runtime::AddressMapping::into_account_id(spender);
+			let amount = Self::u256_to_amount(amount)?;
+
+			ApprovesStorage::<Runtime, Instance>::insert(caller, spender, amount);
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_APPROVAL,
+					context.caller,
+					spender,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	fn transfer(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let to: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		// Build call with origin.
+		{
+			let origin = Runtime::AddressMapping::into_account_id(context.caller);
+			let to = Runtime::AddressMapping::into_account_id(to);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// Dispatch call (if enough gas).
+			let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+				Some(origin).into(),
+				pallet_balances::Call::<Runtime, Instance>::transfer(
+					Runtime::Lookup::unlookup(to),
+					amount,
+				),
+				gasometer.remaining_gas()?,
+			)?;
+			gasometer.record_cost(used_gas)?;
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_TRANSFER,
+					context.caller,
+					to,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	// This function is annotated with transactional.
+	// This is to ensure that if the substrate call fails, the change in allowance is reverted.
+	#[transactional]
```

Comment:
Did you want to leave this annotation now that we know it exists in frontier already?

Reply:
Nice catch ! I'll remove it and check the tests.